controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1292,6 +1292,39 @@ private TxnStatus handleDataNotFoundException(Throwable ex) {
                                 v -> ActiveTxnRecord.parse(v.getValue().getData()))));
     }
 
+    @Override
+    public CompletableFuture<Void> createWaitingRequestIfAbsent(String processorName) {
+        return createWaitingRequestNodeIfAbsent(processorName.getBytes());
+    }
+
+    @Override
+    public CompletableFuture<String> getWaitingRequestProcessor() {
+        return getWaitingRequestNode()
+                .handle((data, e) -> {
+                    if (e != null) {
+                        if (Exceptions.unwrap(e) instanceof DataNotFoundException) {
+                            return null;
+                        } else {
+                            throw new CompletionException(e);
+                        }
+                    } else {
+                        return new String(data.getData());
+                    }
+                });
+    }
+
+    @Override
+    public CompletableFuture<Void> deleteWaitingRequestConditionally(String processorName) {
```

Comment:
There is one optimization that we could do in the whole mechanism to avoid a conditional in two steps. Here is what I'm thinking:

1- Once we create the waiting processor node, we never delete it. To indicate that there is no one waiting, we simply have no processor name.
2- When we write the processor name to the waiting processor node, we keep the version of the node.
3- The delete corresponds to a conditional write to the node using the node version number, and the write is essentially to overwrite the processor name.

This works for zk at least.

Reply:
yes, but it is important to note that WaitingProcRecord is on per-stream basis. It will mean maintaining this state `in-memory` for each stream even if there is no processing forthcoming for the said stream.

Reply:
agreed, there is a space complexity trade-off.